File: Sources/PotatusHub/Localization.swift
import Foundation

/// The menu supports the same bounded language choices as potatoken hub.
/// This stays in Swift rather than `.strings` resources because the release
/// script assembles the app bundle directly from the compiled executable.
enum L {
    enum Language: String, CaseIterable, Sendable {
        case system
        case korean
        case english
        case japanese
        case chinese
    }

    private static let preferenceKey = "PotatusHub.languagePreference"

    static var languagePreference: Language {
        get {
            UserDefaults.standard.string(forKey: preferenceKey)
                .flatMap(Language.init(rawValue:)) ?? .system
        }
        set { UserDefaults.standard.set(newValue.rawValue, forKey: preferenceKey) }
    }

    static var resolved: Language {
        let preference = languagePreference
        guard preference == .system else { return preference }

        let code = Locale.preferredLanguages.first ?? ""
        if code.hasPrefix("ko") { return .korean }
        if code.hasPrefix("ja") { return .japanese }
        if code.hasPrefix("zh") { return .chinese }
        return .english
    }

    static func t(ko: String, en: String, ja: String, zh: String) -> String {
        switch resolved {
        case .korean: return ko
        case .english, .system: return en
        case .japanese: return ja
        case .chinese: return zh
        }
    }
}

File: windows/PotatusHub/Localization.cs
using System.Globalization;

namespace PotatusHub;

public static class L
{
    public enum Language
    {
        System,
        Korean,
        English,
        Japanese,
        Chinese,
    }

    public static Language Preference
    {
        get => AppSettings.Current.Language;
        set
        {
            AppSettings.Current.Language = value;
            AppSettings.Save();
        }
    }

    public static Language Resolved
    {
        get
        {
            if (Preference != Language.System) return Preference;
            var code = CultureInfo.CurrentUICulture.TwoLetterISOLanguageName;
            return code switch
            {
                "ko" => Language.Korean,
                "ja" => Language.Japanese,
                "zh" => Language.Chinese,
                _ => Language.English,
            };
        }
    }

    public static string T(string ko, string en, string ja, string zh) => Resolved switch
    {
        Language.Korean => ko,
        Language.Japanese => ja,
        Language.Chinese => zh,
        _ => en,
    };
}
